mobify/platform-scaffold
# Make Messaging configurable. And great again. Yuge.

Make Messaging configurable.

To support integration of Messaging configuration into the `platform-generator`, add a single `messagingEnabled` value in `package.json` (for both `web` and `native`) that can be set by
the generator code.

 **JIRA**: https://mobify.atlassian.net/browse/MSG-482
 **Linked PRs**: https://github.com/mobify/platform-generator/pull/35

## Changes
- Extend the `webpack` configs `web/webpack/base.loader.js`, `web/webpack/base.main.js` and `native/webpack.app.config.js` to use a `DefinePlugin` to set `MESSAGING_ENABLED` from the `messagingEnabled` value of the appropriate `package.json`. In the case of the app config, also use `MESSAGING_SITE_ID` in the same way that it's done in the web config.
- In `web/app/loader.js`, if `MESSAGING_ENABLED` is false, don't load the PWA Messaging client. Set `window.Progressive.Messaging.enabled` to match `MESSAGING_ENABLED`.
- In `web/app/native/onboarding/onboarding.jsx`, if MESSAGING_ENABLED is false, don't include the `notification` part of the onboarding carousel.
- Set `AstroNative.Configuration.messagingEnabled` from `MESSAGING_ENABLED` in `native/app/config/baseConfig.js`
- In `native/app/onboarding/onboardingModalController.js`, if `AstroNative.Configuration.messagingEnabled` is false, don't initialize the push controller and make the `push:enable` RPC call a no-op

## How to test-drive this PR
- Set `messagingEnabled` to `true` and `false` in `web/package.json` and `native/package.json`> For each setting:
  - in `web`, run `npm run dev`
  - in `native`, run `npm run deps`
  - In Chrome (web) preview and verify that messaging is or is not loaded as expected by checking the console output
  - In XCode, build and run the app and check that the Notifications pane in the carousel is 
included or not as expected
  - In Android Studio, build and run the app and check that the Notifications pane in the carousel is included or not as expected

Note that there are **two** places where `messagingEnabled` is defined, one in each `package.json`. Because of the way code is shared between the `native` and `web` parts of the scaffold, the setting in `web/package.json` affects the PWA **and** the Onboarding part of the app, and the setting in `native/package.json` affects whether the app scaffold initializes messaging.
